keps/sig-scheduling/5075-dra-consumable-capacity/README.md

```diff
@@ -0,0 +1,1256 @@
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+This enhancement introduces the `consumableCapacity` field within the `BasicDevice` of the ResourceSlice. 
+The consumableCapacity can be set to infinity.
+Devices that have at least one consumableCapacity are considered to be shared devices.
+
+Users can only claim a shared device using the newly added `Shared` allocation mode (`DeviceAllocationMode`). In other words, the shared device will be skipped in all other modes, while the unsharable device will be skipped in the Shared mode.
+
+Users can define specific per-device resource requests through
+the newly added resources field in the `DeviceRequest` of the `ResourceClaim`. 
+If the resources field is not specified, all consumable resources are assumed.
+
+A shared device will only be allocated once its consumability has been verified
+and its attributes match the request's selectors and constraints.
+"share" refers to amount of resources reserved or allocated to the claim request.
+The newly added `allocatedShare` field in the `DeviceRequestAllocationResult` will be set when the allocation is successful.
+A device can be shared (allocated) to different pods. 
+However, claims from the same pod cannot allocate shares from the same shared device.
+
+### API enhancement
+To enable this enhancement, the following API updates are proposed.
+#### ResourceSliceSpec's BasicDevice
+
+```go
+type BasicDevice struct {
+	Attributes         map[QualifiedName]DeviceAttribute
+	Capacity           map[QualifiedName]DeviceCapacity
+	ConsumableCapacity map[QualifiedName]DeviceConsumableCapacity
+}
+
+// BasicDevice defines one device instance.
+type BasicDevice struct {
+  ...
+	// ConsumableCapacity defines the set of capacities for this device,
+	// which can be partially consumed and shared by more than one claim.
+	// If at least one item is defined in this field, the device is considered as shared device
+  // and can be allocated only with the `Shared` allocation mode.
+  //
+  // The maximum number of attributes, capacities, and consumable capactites combined is 32.
+  //
+  // +optional
+  // +featureGate=ConsumableCapacity
+  ConsumableCapacity map[QualifiedName]DeviceConsumableCapacity `json:"consumableCapacity,omitempty" protobuf:"bytes,3,rep,name=consumableCapacity"`
+}
+
+// DeviceConsumableCapacity describes a consumable quantity associated with a device.
+type DeviceConsumableCapacity struct {
+  // InfinityResource marks no capacity limit.
+  // If InfinityResource is set to true, Value will be ignored.
+  //
+  // +optional
+  // +default=false
+  InfinityResource bool `json:"infinity" protobuf:"bytes,1,name=infinity"`
```

Comment:
Do we need to advertise capacity if it is infinite? I would assume every device will have some kind of capacity that is finite and I would assume just listing those would be enough.